Implements "completionItem/resolve"

ulugbekna opened the pull request:
Implements `completionItem/resolve` for lazy fetch of docs on code completion.

`textDocument/completion` seems to take now 2-4 times less time. 

Due to the merlin's API, we need to provide it with a document and a position for it to be able to fetch docs. I modify the given document to "fake" the completion. (is there a simpler/cheaper solution?)

I will rebase when we settle on the solution. 

The PR also seems to fix (merlin's?) bug: currently, ocaml-lsp on `Hashtbl.Make<cursor>` shows an empty documentation, but with the new implementation there is documentation.

ulugbekna commented:
`e85c58c` may be controversial and can be removed, but code seems to look less verbose.

rgrinberg commented:
> e85c58c may be controversial and can be removed, but code seems to look less verbose.

It's not controversial but AFAIK, it broke 4.06 last time I tried a change like this.

rgrinberg commented on ocaml-lsp-server/src/ocaml_lsp_server.ml:
```diff
@@ -602,27 +584,39 @@ let on_request :
+  | WillSaveWaitUntilTextDocument _ -> Fiber.return @@ Ok (None, state)
+  | CodeAction params -> code_action rpc params
+  | CompletionItemResolve ({ data; _ } as compl) ->
+    assert (Option.is_some data);
```
Might as well do `let data = Option.value_exn data` instead.

ulugbekna replied:
I do use `Option.value_exn` later, but wanted to document the assumption and thought assert failure would have a more clear stack trace. Removed.

rgrinberg commented on ocaml-lsp-server/src/compl.ml:
```diff
@@ -128,5 +121,56 @@ let complete doc position ~markdown =
+  let compl_info =
+    CompletionParams.(
+      create ~textDocument ~position:lsp_position ()
+      |> yojson_of_t |> Json.t_of_yojson)
```
Is `Json.t_of_yojson` really necessary? It's just the identity function and it exists to just play nice with the code generators.

ulugbekna replied:
You're right. Removed.

rgrinberg commented on ocaml-lsp-server/src/compl.ml:
```diff
@@ -128,5 +121,56 @@ let complete doc position ~markdown =
+    TextDocumentIdentifier.create ~uri:(Document.uri doc |> Lsp.Uri.to_string)
+  in
+  let compl_info =
+    CompletionParams.(
```
Local open should be avoided.

ulugbekna replied:
I thought your comment on local opens to my previous PR was only for `let open ... in`. Here I thought the readability doesn't suffer. Fixed.

rgrinberg commented on ocaml-lsp-server/src/compl.ml:
```diff
@@ -16,7 +16,7 @@ let make_string chars =
   String.init (Array.length chars) ~f:(Array.get chars)
 
-let prefix_of_position source position =
+let prefix_of_position ?(short = false) source position =
```
Let's make this argument mandatory and add a doc string to explain its purpose (or pick a better name perhaps?)

ulugbekna replied:
If `short` is set, the prefix computation for `List.m<cursor>` returns just `m`; otherwise, it returns prefix `List.m`. I first called the param as `short_path` but opted for brevity. Both documenting or giving a more descriptive name (any ideas?) sound good to me.

rgrinberg replied:
Both sound fine. `short_path` sounds clearer to me.

rgrinberg commented:
By the way, I just remembered that `completionItem/resolve` is absurdly slow in Emacs + lsp-mode. I wanted to avoid doing this,  but I don't think it's possible - we need a way to detect the client.

I propose that we introduce a `Client` module:

```ocaml
type t =
  | Emacs
  | Other
```

Make it a part of `State`. Construct an appropriate value from `clientInfo` on initialization. Wdyt?

rgrinberg commented:
Here's the clientInfo from Emacs for reference:

```ocaml
  "clientInfo": {
    "name": "emacs",
    "version": "GNU Emacs 26.3 (build 1, x86_64-apple-darwin19.3.0, Carbon Version 162 AppKit 1894.3)\n of 2020-03-23"
  },
```

ulugbekna commented:
> Make it a part of State. Construct an appropriate value from clientInfo on initialization. Wdyt?

Keeping client info in state sounds like something we would want at some point. A couple of questions:

1. Can't slow `completionItem/resolve` be solved on the `lsp-mode` side? If not, then we could handle docs differently for emacs users. 
2. Should client info code go in this PR?

rgrinberg commented:
> Can't slow completionItem/resolve be solved on the lsp-mode side? If not, then we could handle docs differently for emacs users.

Let me see if this was improved. Last time I asked, I was told it's a limitation of the completion framework in Emacs.

> Should client info code go in this PR?

It can be done separately if you'd like. I don't mind either way.

ulugbekna commented:
Addressed all comments and rebased. 

As per client info, I would prefer for it to go another PR. However, if we don't want to ruin the emacs users' experience with this PR merged before lsp-mode fix lands (if ever), I can think of adding it in this PR but I guess that would mean that docs would be fetched eagerly on emacs.

rgrinberg commented:
> As per client info, I would prefer for it to go another PR. However, if we don't want to ruin the emacs users' experience with this PR merged before lsp-mode fix lands (if ever), I can think of adding it in this PR but I guess that would mean that docs would be fetched eagerly on emacs.

Let's leave the Emacs stuff alone for now actually. I'd like to test things again first as it seems like there were some changes in lsp-mode.

Could you add completionItem/resolve test by the way?

ulugbekna commented:
We need to compute the "suffix" because when we type `List.mp`, the `prefix_of_position ~short_path=true` computes `List.m` instead of `List.mp`, which when applying completion resulted in `List.mapp` passed to merlin.

rgrinberg commented on ocaml-lsp-server/src/compl.ml:
```diff
@@ -31,16 +31,20 @@ let completion_kind kind : CompletionItemKind.t option =
   let chars = Array.of_list chars in
-  String.init (Array.length chars) ~f:(Array.get chars)
+  let arr_len = Array.length chars in
+  (String.init (Array.length chars) ~f:(Array.get chars), arr_len)
```
Why do we return the length here? The caller can always obtain it with `String.length`

ulugbekna replied:
As `make_string` anyway computes the length of the string, I thought why to do that twice.

rgrinberg commented:
String.length is constant time anyways. It’s just a field access
On Jul 9, 2020, 12:02 PM -0700, Ulugbek Abdullaev <notifications@github.com>, wrote:
> @ulugbekna commented on this pull request.
> In ocaml-lsp-server/src/compl.ml:
> > @@ -31,16 +31,20 @@ let completion_kind kind : CompletionItemKind.t option =
>
> let make_string chars =
>   let chars = Array.of_list chars in
> -  String.init (Array.length chars) ~f:(Array.get chars)
> +  let arr_len = Array.length chars in
> +  (String.init (Array.length chars) ~f:(Array.get chars), arr_len)
> As make_string anyway computes the length of the string, I thought why to do that twice.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rgrinberg commented:
Thanks. I made `~short_path` mandatory because there were only 2 callers and squashed some commits. Merging.
